services/src/kilda-pce/src/main/java/org/openkilda/pce/impl/AvailableNetwork.java

```diff
@@ -113,6 +121,43 @@ public void processDiversitySegments(List<PathSegment> segments) {
         }
     }
 
+    /**
+     * Adds diversity weights into {@link AvailableNetwork} based on passed path segments and configuration.
+     */
+    public void processDiversitySegmentsWithPop(List<PathSegment> segments) {
+        if (segments.size() == 1) {
+            return;
+        }
+        int c = 0;
+        Set<String> allocatedPopSet = new HashSet<>();
+
+        for (PathSegment ps : segments) {
+            if (c == 0) {
+                allocatedPopSet.add(ps.getDestSwitch().getPop());
+            } else if (c == segments.size() - 1) {
+                allocatedPopSet.add(ps.getSrcSwitch().getPop());
+            } else {
+                allocatedPopSet.add(ps.getSrcSwitch().getPop());
+                allocatedPopSet.add(ps.getDestSwitch().getPop());
+            }
+            c++;
+        }
+        for (Edge edge : edges) {
+            String srcPop = edge.getSrcSwitch().getPop();
+            if (srcPop != null && allocatedPopSet.contains(srcPop)) {
+                edge.increaseDiversityGroupUseCounter();
+                continue;
+            }
+            String dstPop = edge.getDestSwitch().getPop();
+            if (dstPop != null && allocatedPopSet.contains(dstPop)) {
+                edge.increaseDiversityGroupUseCounter();
+                continue;
+            }
+        }
+
+
+    }
```

Comment:
I think we need some combination of this method and old processDiversitySegments without pops method. Should we try to create diverse path through already used pops if no "unused" pops are available? Maybe add separate PopDiversityGroupUseCounter and related diversity.pop.cost param?

Reply:
I think it makes sense, and overall we should use both strategies at the same time. I introduced new param for config and add new counter for edge, it will also mitigate an issues when there are no PoPs in db yet. good catch